core/sequencer/election.go

```diff
@@ -0,0 +1,133 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sequencer
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/keytransparency/core/domain"
+	"github.com/google/keytransparency/core/mutator"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/util/election"
+)
+
+var isMaster monitoring.Gauge
+
+// Controller runs elections and controls receiver registration for a single domain.
+type Controller struct {
+	d                 *domain.Domain
+	factory           election.Factory
+	maxMastershipHold time.Duration
+}
+
+// NewController creates a new controller.
+func NewController(d *domain.Domain,
+	factory election.Factory,
+	mastershipTimeout time.Duration,
+	mf monitoring.MetricFactory,
+) *Controller {
+	once.Do(func() {
+		isMaster = mf.NewGauge(
+			"master",
+			"Set to 1 for domains this server is master of",
+			domainIDLabel)
+	})
+	return &Controller{
+		d:                 d,
+		factory:           factory,
+		maxMastershipHold: mastershipTimeout,
+	}
+}
+
+// RunWhileMaster waits for mastership before registering a receiver
+// for domain d. During the receiver callback, check for continued mastership.
+func (c *Controller) RunWhileMaster(ctx context.Context,
+	newReceiver func(context.Context, *domain.Domain, func() error) (mutator.Receiver, error)) error {
+
+	el, err := c.factory.NewElection(ctx, c.d.DomainID)
+	if err != nil {
+		return fmt.Errorf("failed to create election for %v: %v", c.d.DomainID, err)
+	}
+	defer func(ctx context.Context) {
+		if err := el.Close(ctx); err != nil {
+			glog.Warningf("%v: Election.Close(): %v", c.d.DomainID, err)
+		}
+	}(ctx)
+	for {
+		if err := c.RunOnce(ctx, el, newReceiver); err != nil {
+			return err
+		}
+	}
+}
+
+// RunOnce waits for mastership, creates a new receiver, and cancels mastership after a bit.
+func (c *Controller) RunOnce(ctx context.Context, el election.MasterElection,
+	newReceiver func(context.Context, *domain.Domain, func() error) (mutator.Receiver, error)) error {
+	defer isMaster.Set(0, c.d.DomainID)
+	isMaster.Set(0, c.d.DomainID)
+	// Wait for mastership.
+	if err := el.WaitForMastership(ctx); err != nil {
+		return err
+	}
+	glog.Infof("Acquired mastership for domain: %v", c.d.DomainID)
+	isMaster.Set(1, c.d.DomainID)
+
+	// Register receiver.
+	// Receiver calls Callback repeatedly in a separate go routine.
+	cancelMastership := make(chan interface{})
+	r, err := newReceiver(ctx, c.d, func() error {
+		return c.callback(ctx, el, cancelMastership)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Wait for resigation.
+	// 10% jitter to avoid many mastership resignations at the same time.
+	jitter := time.Duration(rand.Int63n(int64(c.maxMastershipHold) / 10))
```

Comment:
Note that `rand.Int63n` panics if its argument is 0. Surely it won't normally happen as `maxMastershipHold` is probably more than 9 nanoseconds, but I'd still put a check here.